tensorflow-framework/src/main/java/org/tensorflow/framework/losses/Losses.java

```diff
@@ -0,0 +1,685 @@
+    Operand<T> logcosh = tf.math.sub(tf.math.add(diff, softplus), tf.math.log(two));
+    return tf.math.mean(logcosh, tf.constant(-1));
+  }
+
+  /**
+   * Computes the Poisson loss between labels and predictions.
+   *
+   * <p>The Poisson loss is the mean of the elements of the Tensor <code>
+   * predictions - labels * log(predictions)</code>.
+   *
+   * @param tf the TensorFlow Ops
+   * @param labels true targets
+   * @param predictions the predictions
+   * @param <T> the data type of the predictions and labels
+   * @return the Poisson loss
+   */
+  public static <T extends TNumber, U extends TNumber> Operand<T> poisson(
+      Ops tf, Operand<U> labels, Operand<T> predictions) {
+    DataType<T> dataType = predictions.asOutput().dataType();
+    Operand<T> tLabels = tf.dtypes.cast(labels, dataType);
+    LossTuple<T> lossTuple = LossesImpl.squeezeOrExpandDimensions(tf, tLabels, predictions, null);
+    predictions = lossTuple.getTarget();
+    tLabels = lossTuple.getLabels();
+    Operand<T> epsilonConst = tf.dtypes.cast(tf.constant(EPSILON), dataType);
+
+    return tf.math.mean(
+        tf.math.sub(
+            predictions, tf.math.mul(tLabels, tf.math.log(tf.math.add(predictions, epsilonConst)))),
+        tf.constant(-1));
+  }
+
+  /**
+   * Computes the sparse categorical crossentropy loss between labels and predictions.
+   *
+   * @param tf the TensorFlow Ops
+   * @param labels true targets
+   * @param predictions the predictions
+   * @param fromLogits Whether predictions is expected to be logits. By default, it is assumed that
+   *     predictions encodes a probability distribution.
+   * @param axis The dimension along which the entropy is computed.
+   * @param <T> the data type of the predictions and labels
+   * @return the sparse categorical crossentropy loss
+   */
+  public static <T extends TNumber, U extends TNumber> Operand<T> sparseCategoricalCrossentropy(
+      Ops tf, Operand<U> labels, Operand<T> predictions, boolean fromLogits, int axis) {
+    DataType<T> dataType = predictions.asOutput().dataType();
+    Operand<T> epsilonConst = tf.dtypes.cast(tf.constant(EPSILON), dataType);
+    Operand<T> one = tf.dtypes.cast(tf.constant(1), dataType);
+    Operand<T> oneMinusEpsilonConst = tf.math.sub(one, epsilonConst);
+
+    if (!fromLogits && !(predictions instanceof Variable) && (!tf.scope().env().isEager())) {
+      // TODO output = backtrackIdentity(output); doesn't seem to work with Java version.
+      if (predictions.op().type().equals(Softmax.OP_NAME)) {
+        // When softmax activation function is used for output operation, we
+        // use logits from the softmax function directly to compute loss in order
+        // to prevent collapsing zero when training.
+        // TODO  if( output.op().numOutputs() != 1)
+        //          throw new IllegalArgumentException("output can only have 1 output");
+        // TODO output = output.op.inputs[0]
+        fromLogits = true;
```

Comment:
But it didn't extract the logits, so won't this perform the wrong calculation?

Reply:
Correct, I really need the inputs to the `Softmax` Operation. This is the same issue as mentioned before, 
but I correctly marked these as TODO's

Reply:
Yeah, but I think it would be best to just ignore this optimization and leave the TODO in, rather than half put it in and have it break in the common case.

Reply:
I commented out this code and marked it as a TODO.